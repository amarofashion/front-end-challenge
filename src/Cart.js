import React from 'react';
import Cart from './containers/Cart';
import Header from './containers/Header';
import './index.css';

const App = () => {
    return (
        <div className="mdl-layout mdl-js-layout mdl-color--grey-100">
            <Header />
            <div className="cart-content">
                <Cart />
            </div>
        </div>
    );
}

export default App;
