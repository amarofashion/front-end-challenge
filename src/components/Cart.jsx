import React from 'react';
import { Component } from 'react';
import PropTypes from 'prop-types';

import CartStore from '../stores/Cart.js';

export default class Cart extends Component {
	static propTypes = {
		get: PropTypes.func,
		allProducts: PropTypes.array,
		remove: PropTypes.func
	}

	static defaultProps = {
		get: () => {},
		allProducts: [],
		remove: () => {}
	}

	constructor(props) {
		super(props);
		this.state = {
			products: [],
			showList: false
		};

		this.onChange = this.onChange.bind(this);
		this.toggleList = this.toggleList.bind(this);
		this.remove = this.remove.bind(this);
	}

	componentDidMount() {
		CartStore.listen(this.onChange);
		this.props.get();
	}

	componentWillUnmount() {
		CartStore.unlisten(this.onChange);
	}

	onChange(store) {
		this.setState({
			products: store.products
		});
	}

	toggleList() {
		this.setState({
			showList: !this.state.showList
		});
	}

	remove(evt) {
		this.props.remove(this.state.products[evt.target.dataset.idPosition])
	}

	renderProductsList() {
		const products = this.props.allProducts;
		
		if(!products.length || !this.state.products.length)
			return null;

		return this.state.products.map((id, i) => {
			return <li key={ i } className="product"><a className="link" href={ `#p${id}` }>{ products[id].name }</a><span className="remove" data-id-position={ i } onClick={ this.remove }>X</span></li>
		})
	}

	render() {
		const styleHeight = this.state.showList ? {height: this.state.products.length * 25} : {};
		return (<div className="cart-box">
			<span className="mycart" onClick={ this.toggleList }>Meu carrinho ({ this.state.products.length })</span>
			<ul className={ `products ${this.state.showList ? 'open' : ''}` } style={ styleHeight }>{ this.renderProductsList() }</ul>
		</div>);
	}
}