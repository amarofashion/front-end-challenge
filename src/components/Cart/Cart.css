.cart {
  display: inline-block;
  height: 100%;
  background: #FFF;
  width: 100%;
}

.cart ul {
  border-top: 1px solid #ccc;
  border-bottom: 1px solid #ccc;
  display: inline-block;
  width: 100%;
  padding: 0;
  overflow-y: auto;
  overflow-x: hidden;
  height: 87%;
  position: fixed;
}

.cart h3 {
  padding: 0 20px;
}

.cart ul li {
  list-style: none;
  padding: 10px;
  border-bottom: 1px solid #ccc;
}

.cart ul li:last-child {
  border: none;
}
.cart-item {
  margin-bottom: .3em;
  position: relative;
  display: inline-block;
  width: 100%;
}

.cart-item__qtd,
.cart-item__price,
.cart-item__name {
  display: block;
}

.cart__total {
  font-weight: bold;
  font-size: 1.5em;
  line-height: 1.1em; 
  text-align: right;
  padding: 10px;
  background-color: #000;
  color: #FFF;
  position: fixed;
  width: 100%;
  bottom: 0;
  left: 0;
}

.cart__total p {
  padding: 10px;
  margin: 0 20px 0 0;
  font-size: 20px;
}

.btn-close {
  font-size: 12px;
  background: none;
  border: none;
  position: absolute;
  top: 0;
  right: 0;
}

.cart-item .cart-thumbnail {
  width: 15%;
  float: left;
} 

.empty-cart {
  padding: 40px;
  text-align: center;
}

.empty-cart h3 {
  padding: 0;
}

.bt-buy-more {
  margin: 0 auto;
  border: solid 1px #000;
  padding: 10px;
  display: inline-flex;
  color: #000;
  background-color: #FFF;
}

.bt-buy-more:hover {
  text-decoration: none;
  color: #000;
}

.cart-details {
  float: left;
  width: 70%;
  padding: 10px;
}

.description {
  padding: 15px;
  min-height: 70px;;
}

.item-description {
  width: 100%;
}

.cart-container {
  width: 100%;
}