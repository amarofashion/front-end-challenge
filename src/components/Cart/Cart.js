import React from 'react';
import { Link } from 'react-router-dom';
import PropTypes from 'prop-types';
import CartItem from '../../containers/CartItem';

const Cart = ({ items, total, currency, quantity }) => {
    return (
        <div className="cart container">
            <h3>Sacola ({ quantity })</h3>
            <div className="cart-container">
                {items.length > 0 && (
                    <ul>
                        {items.map(item => (
                            <li key={item.id}>
                                <CartItem {...item} /> 
                            </li>
                        ))}
                    </ul>
                )}
                {items.length === 0 && (
                    <div className="empty-cart">
                        <h3>SUA SACOLA EST√Å VAZIA :(</h3>
                        <Link className="bt-buy-more" title="Comprar mais" to="/">Comprar mais</Link>
                    </div>
                )}
            </div>
            <div className="cart__total">
                <p>Total: {currency} {total}</p>
            </div>
        </div>
    );
}

Cart.propTypes = {
    items: PropTypes.array,
    total: PropTypes.string,
    currency: PropTypes.string,
    quantity: PropTypes.number
}

export default Cart;
