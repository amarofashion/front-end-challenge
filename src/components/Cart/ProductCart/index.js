import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Thumb from './../../Thumb';

class ProductCart extends Component {
  static propTypes = {
    product: PropTypes.object.isRequired,
    removeProduct: PropTypes.func.isRequired,
    changeProductQuantity: PropTypes.func.isRequired,
  };

  constructor(props) {
    super(props);
    this.state = {
      product: this.props.product
    };
  }

  handlePlus = () => {
    const { changeProductQuantity } = this.props;
    const { product } = this.state;
    product.quantity = product.quantity + 1;
    changeProductQuantity(product);
  }

  handleMinus = () => {
    const { changeProductQuantity } = this.props;
    const { product } = this.state;
    product.quantity = product.quantity - 1;
    changeProductQuantity(product);
  }

  render() {
    const { removeProduct } = this.props;
    const { product } = this.state;

    return (
      <div className="list-item">
        <div
          className="list-item__del"
          onClick={() => removeProduct(product)}
        >
            Remover
        </div>
        <Thumb
          classes="list-item__thumb"
          src={product.image}
          alt={product.name}
        />
        <div className="list-item__details">
          <p className="title">{product.name}</p>
          <p className="desc">
            {`${product.variant}`} <br />
            Quantidade: {product.quantity}
          </p>
        </div>
        <div className="list-item__price">
          <p>{product.actual_price}</p>
          <div>
            <button onClick={this.handleMinus} disabled={product.quantity === 1 ? true : false} className="change-product-button">-</button>
            <button onClick={this.handlePlus} className="change-product-button">+</button>
          </div>
        </div>
      </div>
    );
  }
}

export default ProductCart;
