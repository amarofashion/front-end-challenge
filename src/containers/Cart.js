import { connect } from 'react-redux'
import CartList from '../components/CartList'
import { removeFromCart, increment, decrement } from '../actions'

const mapStateToProps = state => {
  return { items: state.cart }
}

const mapDispatchToProps = dispatch => {
  return {
    onRemove(id) {
      dispatch(removeFromCart(id))
    },
    onIncrement(id) {
      dispatch(increment(id))
    },
    onDecrement(id) {
      dispatch(decrement(id))
    }
  }
}

const Cart = connect(
  mapStateToProps,
  mapDispatchToProps
)(CartList)

export default Cart