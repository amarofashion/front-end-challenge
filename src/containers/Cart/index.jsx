import React, { Component } from 'react';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import { withRouter } from 'react-router';
import styled from 'styled-components';
import CartTile from '../../components/CartTile';
import * as CartActions from '../../actions/cart';

const CartWrapper = styled.ul`
    display: flex;
    flex-wrap: wrap;
    flex-direction: column;
    margin-top: 75px;
`

const EndButton = styled.button`
    height: 50px;
    background: black;
    font-size: 14px;
    color: white;
    width: 320px;
    margin: 8px;
`;

class Cart extends Component {
    render() {
        const { cart } = this.props.state;

        return(
            <CartWrapper>
                {cart.items.map((product, key) => (
                    <CartTile
                        key={key}
                        product={product}
                    />
                ))}

                <EndButton>{cart.total} Comprar</EndButton>
            </CartWrapper>
        );
    }
}

export default withRouter(connect(
    state => ({ state }),
    dispatch => ({
        actions: bindActionCreators({
            ...CartActions,
        }, dispatch)
    })
)(Cart));
