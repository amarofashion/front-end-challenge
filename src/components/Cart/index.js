import React, { Component } from "react";
import PropTypes from "prop-types";
import { MdShoppingCart } from "react-icons/md";
import { connect } from "react-redux";
import {
  loadCart,
  removeProduct,
  changeProductQuantity
} from "../../api/redux/reducers/cart/actions";
import { updateCart } from "../../api/redux/reducers/total/actions";
import ProductCart from "./ProductCart";

import "./style.scss";

class Cart extends Component {
  static propTypes = {
    loadCart: PropTypes.func.isRequired,
    updateCart: PropTypes.func.isRequired,
    cartProducts: PropTypes.array.isRequired,
    newProduct: PropTypes.object,
    removeProduct: PropTypes.func,
    productToRemove: PropTypes.object,
    changeProductQuantity: PropTypes.func,
    productToChange: PropTypes.object
  };

  state = {
    isOpen: false
  };

  componentWillReceiveProps(nextProps) {
    if (nextProps.newProduct !== this.props.newProduct) {
      this.addProduct(nextProps.newProduct);
    }

    if (nextProps.productToRemove !== this.props.productToRemove) {
      this.removeProduct(nextProps.productToRemove);
    }

    if (nextProps.productToChange !== this.props.productToChange) {
      this.changeProductQuantity(nextProps.productToChange);
    }
  }

  openCart = () => {
    this.setState({ isOpen: true });
  };

  closeCart = () => {
    this.setState({ isOpen: false });
  };

  addProduct = product => {
    const { cartProducts, updateCart } = this.props;
    let productAlreadyInCart = false;

    cartProducts.forEach(cp => {
      if (cp.id === product.id) {
        cp.quantity += product.quantity;
        productAlreadyInCart = true;
      }
    });

    if (!productAlreadyInCart) {
      cartProducts.push(product);
    }

    updateCart(cartProducts);
    this.openCart();
  };

  removeProduct = product => {
    const { cartProducts, updateCart } = this.props;

    const index = cartProducts.findIndex(p => p.id === product.id);
    if (index >= 0) {
      cartProducts.splice(index, 1);
      updateCart(cartProducts);
    }
  };

  changeProductQuantity = changedProduct => {
    const { cartProducts, updateCart } = this.props;

    const product = cartProducts.find(p => p.id === changedProduct.id);
    product.quantity = changedProduct.quantity;
    if (product.quantity <= 0) {
      this.removeProduct(product);
    }
    updateCart(cartProducts);
  };

  render() {
    const {
      cartTotal,
      cartProducts,
      removeProduct,
      changeProductQuantity
    } = this.props;

    const products = cartProducts.map(p => {
      return (
        <ProductCart
          product={p}
          removeProduct={removeProduct}
          changeProductQuantity={changeProductQuantity}
          key={p.id}
        />
      );
    });

    let classes = ["cart"];

    if (!!this.state.isOpen) {
      classes.push("cart--open");
    }

    return (
      <div className={classes.join(" ")}>
        {this.state.isOpen && (
          <div
            onClick={() => this.closeCart()}
            className="cart__close-btn"
          >
            X
          </div>
        )}
        {!this.state.isOpen && (
          <span
            onClick={() => this.openCart()}
            className="bag bag--cart-closed"
          >
            <MdShoppingCart/>
            <span className="bag__quantity">{cartTotal.productQuantity}</span>
          </span>
        )}

        <div className="cart__content">
          <div className="cart__header">
            <span className="header-title">Produtos {cartTotal.productQuantity}</span>
          </div>

          <div className="cart__list-container">
            {products}
            {!products.length && (
              <p className="list-empty">
                Carrinho Vazio<br />
                :(
              </p>
            )}
          </div>

          <div className="cart__footer">
            <div className="sub">SUBTOTAL</div>
            <div className="sub-price">
              <p className="sub-price__val">{cartTotal.totalPrice.toFixed(2)}</p>
            </div>
          </div>
        </div>
      </div>
    );
  }
}

const mapStateToProps = state => ({
  cartProducts: state.cart.products,
  newProduct: state.cart.productToAdd,
  productToRemove: state.cart.productToRemove,
  productToChange: state.cart.productToChange,
  cartTotal: state.total.data
});

export default connect(
  mapStateToProps,
  { loadCart, updateCart, removeProduct, changeProductQuantity }
)(Cart);
