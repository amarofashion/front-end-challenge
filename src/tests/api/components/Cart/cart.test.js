import React from 'react';
import Cart from '../../../../components/Cart/index';
import ProductCart from '../../../../components/Cart/ProductCart/index';
import renderer from 'react-test-renderer';
import CustomProvider from '../../../../../src/provider/index';

const dataProduct = {
    name: "CASACO WHITE FUR",
    style: "20002648",
    sku: "5891_1000058_0_PP",
    code_color: "20002648_029",
    color_slug: "off-white",
    color: "OFF-WHITE",
    on_sale: false,
    regular_price: "R$ 239,90",
    actual_price: "R$ 239,90",
    variant: 'PP',
    discount_percentage: "",
    installments: "3x R$ 79,97",
    image: "https://d3l7rqep7l31az.cloudfront.net/images/products/20002648_029_catalog_1.jpg?1459552552",
};


describe("Cart Component", () => {
  test("Wrapper cart component test", () => {
    const tree = renderer.create( 
    <CustomProvider>
       <Cart />
    </CustomProvider>).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Wrapper product cart component test", () => {
    const tree = renderer.create( 
    <CustomProvider>
       <ProductCart product={dataProduct} 
        changeProductQuantity={() => console.log('run')} 
        removeProduct ={() => console.log('run')}  />
    </CustomProvider>).toJSON();
    expect(tree).toMatchSnapshot();
  });
});