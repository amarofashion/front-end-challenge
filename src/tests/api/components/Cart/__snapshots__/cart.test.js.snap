// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Component Wrapper cart component test 1`] = `
<div
  className="cart"
>
  <span
    className="bag bag--cart-closed"
    onClick={[Function]}
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M7 18c-1.1 0-1.99.9-1.99 2S5.9 22 7 22s2-.9 2-2-.9-2-2-2zM1 2v2h2l3.6 7.59-1.35 2.45c-.16.28-.25.61-.25.96 0 1.1.9 2 2 2h12v-2H7.42c-.14 0-.25-.11-.25-.25l.03-.12.9-1.63h7.45c.75 0 1.41-.41 1.75-1.03l3.58-6.49c.08-.14.12-.31.12-.48 0-.55-.45-1-1-1H5.21l-.94-2H1zm16 16c-1.1 0-1.99.9-1.99 2s.89 2 1.99 2 2-.9 2-2-.9-2-2-2z"
      />
    </svg>
    <span
      className="bag__quantity"
    >
      0
    </span>
  </span>
  <div
    className="cart__content"
  >
    <div
      className="cart__header"
    >
      <span
        className="header-title"
      >
        Produtos 
        0
      </span>
    </div>
    <div
      className="cart__list-container"
    >
      <p
        className="list-empty"
      >
        Carrinho Vazio
        <br />
        :(
      </p>
    </div>
    <div
      className="cart__footer"
    >
      <div
        className="sub"
      >
        SUBTOTAL
      </div>
      <div
        className="sub-price"
      >
        <p
          className="sub-price__val"
        >
          0.00
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Cart Component Wrapper product cart component test 1`] = `
<div
  className="list-item"
>
  <div
    className="list-item__del"
    onClick={[Function]}
  >
    Remover
  </div>
  <div
    className="list-item__thumb"
  >
    <img
      alt="CASACO WHITE FUR"
      src="https://d3l7rqep7l31az.cloudfront.net/images/products/20002648_029_catalog_1.jpg?1459552552"
    />
  </div>
  <div
    className="list-item__details"
  >
    <p
      className="title"
    >
      CASACO WHITE FUR
    </p>
    <p
      className="desc"
    >
      PP
       
      <br />
      Quantidade: 
    </p>
  </div>
  <div
    className="list-item__price"
  >
    <p>
      R$ 239,90
    </p>
    <div>
      <button
        className="change-product-button"
        disabled={false}
        onClick={[Function]}
      >
        -
      </button>
      <button
        className="change-product-button"
        onClick={[Function]}
      >
        +
      </button>
    </div>
  </div>
</div>
`;
